hail/src/main/scala/is/hail/annotations/aggregators/RegionValueApproxCDFAggregator.scala

```diff
@@ -0,0 +1,706 @@
+ *   The values in level i occupy indices from `levels(i)` (inclusive) to
+ *   `levels(i+1)` (exclusive).
+ * - `numLevels` is the number of levels currently held. The top level is
+ *   never empty, so this is also the greatest nonempty level.
+ */
+class ApproxCDFCombiner[@specialized(Int, Long, Float, Double) T: ClassTag : Ordering](
+  val levels: Array[Int], val items: Array[T], var numLevels: Int, var minValue: T, var maxValue: T
+)(implicit helper: ApproxCDFHelper[T]
+) extends Serializable {
+  def this(levels: Array[Int], items: Array[T], numLevels: Int)(implicit helper: ApproxCDFHelper[T]) =
+    this(levels, items, numLevels, helper.dummyValue, helper.dummyValue)
+
+  def copy(): ApproxCDFCombiner[T] =
+    new ApproxCDFCombiner[T](levels.clone(), items.clone(), numLevels, minValue, maxValue)
+
+  def maxNumLevels = levels.length - 1
+  def capacity = items.length
+  def isFull = levels(0) == 0
+
+  def size = levels(maxNumLevels) - levels(0)
+
+  def levelSize(level: Int): Int = levels(level + 1) - levels(level)
+
+  def safeLevelSize(level: Int): Int =
+    if (level >= maxNumLevels) 0 else levels(level + 1) - levels(level)
+
+  def push(t: T) {
+    val bot = levels(0)
+
+    if (bot == capacity) {
+      minValue = t
+      maxValue = t
+    } else if (helper.lt(t, minValue)) {
+      minValue = t
+    } else if (helper.lt(maxValue, t)) {
+      maxValue = t
+    }
+
+    val newBot = bot - 1
+    items(newBot) = t
+    levels(0) = newBot
+
+  }
+
+  def grow(newNumLevels: Int, newCapacity: Int): ApproxCDFCombiner[T] = {
+    require(newNumLevels > maxNumLevels && newCapacity > capacity)
+    val newLevels = Array.ofDim[Int](newNumLevels + 1)
+    val newItems = Array.ofDim[T](newCapacity)
+    val shift = newCapacity - capacity
+    var i = 0
+    while (i <= maxNumLevels) {
+      newLevels(i) = levels(i) + shift
+      i += 1
+    }
+    while (i <= newNumLevels) {
+      newLevels(i) = newCapacity
+      i += 1
+    }
+    System.arraycopy(items, levels(0), newItems, newLevels(0), size)
+
+    new ApproxCDFCombiner[T](newLevels, newItems, numLevels, minValue, maxValue)
+  }
+
+  def clear() {
+    minValue = helper.dummyValue
+    maxValue = helper.dummyValue
+    numLevels = 1
+    var i = 0
+    while (i < levels.length) {
+      levels(i) = items.length
+      i += 1
+    }
+  }
+
+  /* Compact level `level`, merging the compacted results into level `level+1`.
+   * Shift lower levels up to keep items contiguous.
+   */
+  def compactLevel(level: Int, shiftLowerLevels: Boolean = true) {
+    assert(level < maxNumLevels - 1)
+    if (level == numLevels - 1) numLevels += 1
+
+    val bot = levels(0)
+    val a0 = levels(level)
+    val c = levels(level + 1)
+    val d = levels(level + 2)
+
+    val size = c - a0
+    val halfSize = size / 2
+    val adj = size % 2
+
+    val a = a0 + adj
+    val b = a + halfSize
+
+    val sizeBelow = a - bot
+    val sizeAbove = d - c
```

Comment:
there is a distressing asymmetry between level 0 to level and level+1 to level+2 (rather than levelMax).